File: src/BjjEire.Application/Features/BjjEvents/DTOs/BjjEventDto.cs
using BjjEire.Application.Common;
using BjjEire.Application.Common.DTOs;
using BjjEire.Domain.Entities.BjjEvents;
using BjjEire.Domain.Enums;

namespace BjjEire.Application.Features.BjjEvents.DTOs;

public class BjjEventDto : BaseApiEntityModel
{
    public string Name { get; set; } = string.Empty;

    public string? Description { get; set; } = string.Empty;

    public BjjEventType Type { get; set; }

    public OrganizerDto Organiser { get; set; } = new();

    public EventStatus Status { get; set; }

    public string? StatusReason { get; set; }

    public SocialMediaDto SocialMedia { get; set; } = new();

    public County County { get; set; }

    public LocationDto Location { get; set; } = new();

    public BjjEventScheduleDto Schedule { get; set; } = null!;

    public PricingModelDto Pricing { get; set; } = new();

    public string EventUrl { get; set; } = string.Empty;

    public string ImageUrl { get; set; } = string.Empty;

}

public class BjjEventMapping : Profile
{
    public BjjEventMapping()
    {
        _ = CreateMap<BjjEvent, BjjEventDto>()
          .ForMember(dest => dest.CreatedOnUtc, mo => mo.Ignore())
          .ForMember(dest => dest.UpdatedOnUtc, mo => mo.Ignore());
        _ = CreateMap<BjjEventDto, BjjEvent>()
            .ForMember(dest => dest.CreatedOnUtc, mo => mo.Ignore())
            .ForMember(dest => dest.UpdatedOnUtc, mo => mo.Ignore());
    }
}

File: src/bjjeire-api/src/main/java/com/bjjeire/api/event/BjjEventDto.java
package com.bjjeire.api.event;

import com.bjjeire.api.common.County;
import com.bjjeire.api.common.Location;
import com.bjjeire.api.common.SocialMedia;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.List;

@ValidBjjEvent
public record BjjEventDto(
        String id,
        String name,
        String description,
        List<BjjEventType> types,
        Organizer organiser,
        EventStatus status,
        String statusReason,
        SocialMedia socialMedia,
        County county,
        Location location,
        BjjEventSchedule schedule,
        List<PricingModel> pricingOptions,
        String eventUrl,
        String imageUrl,
        boolean isActive,

        @JsonProperty(access = JsonProperty.Access.READ_ONLY)
        List<CalculatedCost> calculatedCosts) {

    public BjjEventDto(
            String id,
            String name,
            String description,
            List<BjjEventType> types,
            Organizer organiser,
            EventStatus status,
            String statusReason,
            SocialMedia socialMedia,
            County county,
            Location location,
            BjjEventSchedule schedule,
            List<PricingModel> pricingOptions,
            String eventUrl,
            String imageUrl,
            boolean isActive) {
        this(
                id,
                name,
                description,
                types,
                organiser,
                status,
                statusReason,
                socialMedia,
                county,
                location,
                schedule,
                pricingOptions,
                eventUrl,
                imageUrl,
                isActive,
                null);
    }
}
